resources/user.go

```diff
@@ -0,0 +1,305 @@
+		}
+	}
+
+	if obj.AllowDuplicateUID == false && obj.UID != nil {
+		existingUID, err := user.LookupId(strconv.Itoa(int(*obj.UID)))
+		if err == nil {
+			if existingUID.Username != obj.GetName() {
+				return false, fmt.Errorf("the uid: %d is already taken", *obj.UID)
+			} else {
+				return false, errwrap.Wrapf(err, "error looking up uid")
+			}
+		}
+	}
+
+	if usercheck := true; exists && obj.State == "exists" {
+		if obj.GID != nil {
+			if strconv.Itoa(int(*obj.GID)) != usr.Gid {
+				usercheck = false
+			}
+		}
+		if obj.UID != nil {
+			if strconv.Itoa(int(*obj.UID)) != usr.Uid {
+				usercheck = false
+			}
+		}
+		if obj.HomeDir != nil {
+			if *obj.HomeDir != usr.HomeDir {
+				usercheck = false
+			}
+		}
+		if usercheck {
+			return true, nil
+		}
+	}
+
+	if obj.State == "absent" && !exists {
+		return true, nil
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	var cmdName string
+	var args []string
+	if obj.State == "exists" {
+		if !exists {
+			cmdName = "useradd"
+			log.Printf("adding user: %s", obj.GetName())
+		}
+		if exists {
+			cmdName = "usermod"
+			log.Printf("modifying user: %s", obj.GetName())
+		}
+		if obj.AllowDuplicateUID == true {
+			args = append(args, "--non-unique")
+		}
+		if obj.UID != nil {
+			args = append(args, "-u", fmt.Sprintf("%d", *obj.UID))
+		}
+		if obj.GID != nil {
+			args = append(args, "-g", fmt.Sprintf("%d", *obj.GID))
+		}
+		if obj.HomeDir != nil {
+			args = append(args, "-d", *obj.HomeDir)
+		}
+
+	}
+
+	if obj.State == "absent" {
+		cmdName = "userdel"
+	}
+
+	args = append(args, obj.GetName())
+
+	cmd := exec.Command(cmdName, args...)
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+		Pgid:    0,
+	}
+	if err := cmd.Run(); err != nil {
+		return false, errwrap.Wrapf(err, "cmd failed to run")
+	}
+
+	return false, nil
+}
+
+// UserUID is the UID struct for UserRes.
+type UserUID struct {
+	BaseUID
+	name string
+}
+
+// UIDs includes all params to make a unique identification of this object.
+// Most resources only return one, although some resources can return multiple.
+func (obj *UserRes) UIDs() []ResUID {
+	x := &UserUID{
+		BaseUID: BaseUID{Name: obj.GetName(), Kind: obj.GetKind()},
+		name:    obj.Name,
+	}
+	return []ResUID{x}
+}
+
+// GroupCmp returns whether two resources can be grouped together or not.
+func (obj *UserRes) GroupCmp(r Res) bool {
+	_, ok := r.(*UserRes)
+	if !ok {
+		return false
+	}
+	return false
+}
+
+// Compare two resources and return if they are equivalent.
+func (obj *UserRes) Compare(r Res) bool {
+	// we can only compare UserRes to others of the same resource kind
+	res, ok := r.(*UserRes)
+	if !ok {
+		return false
+	}
+	if !obj.BaseRes.Compare(res) { // call base Compare
+		return false
+	}
+	if obj.Name != res.Name {
+		return false
+	}
+	if obj.State != res.State {
+		return false
+	}
+	if obj.AllowDuplicateUID != res.AllowDuplicateUID {
+		return false
+	}
+	if obj.UID != nil && res.UID != nil {
```

Comment:
what happens if obj.UID is nil and res.UID is not nil?

the clean solution to this can be found in file.go